### DISCUSSION TITLE
Segmentation with tsai

### DISCUSSION THREAD

**Discussion by carbocation** [timeseriesAI/tsai#61] [Q&A]
Is there a concept of segmentation in `tsai`? E.g., I have a time series and for every single point, one label applies. This is a 1-dimensional equivalent of semantic segmentation. And I'd like to train a model to perform this same task. I didn't see an obvious candidate with `tsai` but am curious if there is an example that seems to fit.

**Comment by vrodriguezf** [comment]
Not that I am aware of. Something like U-net for time series was proposed at NeurIPS a copule of years ago:
https://arxiv.org/abs/1910.11162

**Comment by carbocation** [comment]
Thanks for the pointer. It looks like there is an [implementation in <s>PyTorch</s> TensorFlow](https://github.com/perslev/U-Time). I'll have to dig into this.

One that is actually in [PyTorch is here](https://github.com/mtdo/ts-segment) and their FCN [model for semantic segmentation is here](https://github.com/mtdo/ts-segment/blob/master/segmentation/models/models.py).

**Reply by vrodriguezf** [reply]
There's also one here:
https://github.com/wangjinzhuo/wearables/blob/master/utime/utime.py

**Comment by oguiza** [comment]
Since this is more a (very interesting) question than an issue, I'll move it to discussions.

**Reply by carbocation** [reply]
Thank you! (I was actually not previously aware that github had discussions - very cool.)

**Comment by oguiza** [comment]
Hi,
That's a very interesting question James (@carbocation). I hadn't seen any time series segmentation task, but it's very interesting and makes a lot of sense. 
And thanks for the paper Victor (@vrodriguezf). It's very helpful!
Are any of you you aware of any public time-series segmentation dataset? I'd like to better understand the target shape. 
There's nothing related to segmentation, but I'd be interested in adding it.
It'd be good if we update each other with our findings on this, with the idea of adding a segmentation model to `tsai`.

**Reply by carbocation** [reply]
I actually don't know of a pre-existing public dataset for this exact task. For a related task (point-by-point regression between two paired time series - which is conceptually similar but with a continuous output per point instead of a categorical output per point) the [PhysioNet data](https://physionet.org/content/ecgiddb/1.0.0/) might work. 

In that dataset, there is a paired unfiltered ECG signal and then a filtered ECG signal. Of course it's not that interesting - the model would be learning a basic filter - but it's public and paired which I think are the key features. 

(I consider this related because it has the same property where the output duration is the same as the input duration and every input data point needs its own output data point. But people might not agree that it's very similar.)

**Reply by carbocation** [reply]
And perhaps a more direct segmentation example is available from [ts-segment](https://github.com/mtdo/ts-segment) which @vrodriguezf identified. That is the MobiActV2 dataset, which has actual class labels for different periods of time.
